File: solutions/day11_part1.py
with open("./input.txt", "r") as inp:
  sto: str = inp.read().rstrip()
  inp.close()

stones = list(map(int, sto.split(' ')))
print(stones)

for i in range(25):
    j = 0
    new_stones = []
    while j < len(stones):
        if stones[j] == 0:
            new_stones.append(1)
            # stones[j] = 1
        elif len(str(stones[j])) % 2 == 0:
            stone = str(stones[j])
            new_stones.append(int(stone[:len(stone) // 2]))
            new_stones.append(int(stone[len(stone) // 2:]))
            # stones = stones[:j] + [int(stone[:len(stone) // 2])] + [int(stone[len(stone) // 2:])] + stones[j + 1:]
            # j += 1
        else:
            new_stones.append(stones[j] * 2024)
            # stones[j] *= 2024
        j += 1
    stones = new_stones

print(len(stones))

File: solutions/day11_part1.ts
import fs from "node:fs";
import { LRUCache } from "npm:lru-cache@11.0.2";

const raw = fs.readFileSync("./input.txt", "utf-8");
const stones = raw.split(" ").map(Number);

function splitStone(stone: number): number[] {
  // const stringifiedStone = stone.toString();
  // const midIndex = stringifiedStone.length / 2;
  //   console.log(
  //     "setting",
  //     Array.from(stonesMap.set(stone, splittedStone)).pop()
  //   );
  const denominator = 10 ** ((Math.floor(Math.log10(stone)) - 1) / 2 + 1);
  const firstHalf = Math.floor(stone / denominator);
  const secondHalf = stone % denominator;
  return [firstHalf, secondHalf];
}

const options = {
  max: 1000,
};
function iterateStones(
  stones: number[],
  iterationLeft: number,
  stonesMap: LRUCache<number, number[]> = new LRUCache(options)
): number[] {
  // console.log("iteration", iterationLeft);
  if (iterationLeft === 0) return stones;
  const newStones = stones.flatMap((stone) => {
    if (stone === 0) {
      return [1];
    } else if (Math.floor(Math.log10(stone)) % 2 === 1) {
      const foundStone = stonesMap.get(stone);
      if (foundStone) {
        return foundStone;
      } else {
        const splittedStone = splitStone(stone);
        stonesMap.set(stone, splittedStone);
        return splittedStone;
      }
    } else {
      return [stone * 2024];
    }
  });
  return iterateStones(newStones, iterationLeft - 1, stonesMap);
}

const result = iterateStones(stones, 25);
console.log(result.length);
